com.unity.shadergraph/Editor/Generation/Processors/Generator.cs

```diff
@@ -190,53 +250,93 @@ void BuildShader()
             // Collect excess shader properties from the TargetImplementation
             foreach (var target in m_Targets)
             {
-                // TODO: Setup is required to ensure all Targets are initialized
-                // TODO: Find a way to only require this once
-                TargetSetupContext context = new TargetSetupContext();
-                target.Setup(ref context);
-
                 target.CollectShaderProperties(shaderProperties, m_Mode);
             }
 
             // set the property collector to read only
             // (to ensure no rogue target or pass starts adding more properties later..)
             shaderProperties.SetReadOnly();
 
-            m_Builder.AppendLine(@"Shader ""{0}""", m_Name);
+            m_Builder.AppendLine(@"Shader ""{0}""", shaderName);
             using (m_Builder.BlockScope())
             {
+                var shaderDependencies = new List<ShaderDependency>();
+                var shaderCustomEditors = new List<ShaderCustomEditor>();
+                string shaderCustomEditor = typeof(GenericShaderGraphMaterialGUI).FullName;
+                string shaderFallback = "Hidden/Shader Graph/FallbackError";
+
                 GenerationUtils.GeneratePropertiesBlock(m_Builder, shaderProperties, shaderKeywords, m_Mode, graphInputOrderData);
-                for (int i = 0; i < m_Targets.Length; i++)
+                for (int i = 0; i < m_Targets.Count; i++)
                 {
-                    TargetSetupContext context = new TargetSetupContext(m_assetCollection);
+                    var context = m_TargetContexts[i];
 
-                    // Instead of setup target, we can also just do get context
-                    m_Targets[i].Setup(ref context);
-
-                    var subShaderProperties = GetSubShaderPropertiesForTarget(m_Targets[i], m_GraphData, m_Mode, m_OutputNode, m_TemporaryBlocks);
-                    foreach (var subShader in context.subShaders)
+                    // process the subshaders
+                    var subShaderProperties = GetSubShaderPropertiesForTarget(m_Targets[i], m_GraphData, m_Mode, m_OutputNode, isPrimaryShader ? m_PrimaryShaderTemporaryBlocks : null);
```

Comment:
nit: If `m_PrimaryShaderTemporaryBlocks` is used in all code paths, is it really only for the primary shader?

Reply:
it's only passed down here (to fill out the data structure) for the primary shader.

Reply:
But I can probably make this even more obvious by just passing in the temporary block list as a parameter to BuildShader()

Reply:
oh I think i see, didn't notice the turnary